File: frontend/src/lib/services/localIndicators.ts
/**
 * Query options for the LCFF Local Indicator API.
 *
 * This is the local half of the Dashboard, and it works nothing like the
 * state half in `accountability.ts`. There are no performance colours here —
 * only `Met`, `Not Met` or `Not Met For Two or More Years` — because nothing
 * on this side is measured by the state. Each local educational agency rates
 * itself and reports the result to its own governing board.
 */
import {
	localIndicatorsReadCatalogOptions,
	localIndicatorsReadLocalIndicatorsOptions,
	localIndicatorsReadPriorityOptions,
	localIndicatorsReadTrendOptions,
} from '@/lib/client'
import { reference, report } from '@/lib/services/query'

export const STATEWIDE_CDS = '00000000000000'

export function localIndicatorCatalogQuery(year?: number) {
	return reference(
		localIndicatorsReadCatalogOptions({ query: year ? { year } : {} }),
		'Could not load the local indicator catalogue.',
	)
}

export function localIndicatorsQuery(cds: string, year: number) {
	return report(
		localIndicatorsReadLocalIndicatorsOptions({ query: { cds, year } }),
		'Could not load the local indicators.',
	)
}

export function localIndicatorDetailQuery(cds: string, year: number, priority: number) {
	return report(
		localIndicatorsReadPriorityOptions({ query: { cds, year, priority } }),
		'Could not load that priority.',
	)
}

export function localIndicatorTrendQuery(cds: string, priority: number) {
	return report(
		localIndicatorsReadTrendOptions({ query: { cds, priority } }),
		'Could not load the history for that priority.',
	)
}

/**
 * How a `Met` / `Not Met` value should read.
 *
 * Deliberately not the Dashboard's five colours: this side of the Dashboard
 * has no performance colour, and borrowing the palette would imply the state
 * measured something it did not.
 */
export function performanceTone(
	performance: string | null | undefined,
): 'met' | 'notMet' | 'notMetTwoYears' | 'none' {
	switch (performance) {
		case 'Met':
			return 'met'
		case 'Not Met':
			return 'notMet'
		case 'Not Met For Two or More Years':
			return 'notMetTwoYears'
		default:
			return 'none'
	}
}

/** Turn a published column name into something a reader can scan. */
export function humanizeField(field: string): string {
	return field
		.replace(/^Narrative/, 'Narrative ')
		.replace(/([a-z])([A-Z])/g, '$1 $2')
		.replace(/\s+/g, ' ')
		.trim()
		.replace(/^./, (c) => c.toUpperCase())
}

File: backend/app/model/local_indicators.py
"""The LCFF Local Indicators -- the other half of the California Dashboard.

The Dashboard has two halves.  :mod:`app.model.dashboard` holds the state
half: seven indicators the state measures, each reduced to a performance
colour.  This module holds the local half, which works nothing like it.

Local Indicators are **self-assessments**.  Each local educational agency
rates itself against the LCFF state priorities, reports the result to its own
governing board at a public meeting, and the state records what it said.  There
is no colour, no five-by-five grid and no cut points; there is only ``Met``,
``Not Met``, or ``Not Met For Two or More Years``, plus whatever the LEA chose
to write.

Three consequences shape the schema:

``performance`` is a string, never a colour
    Nothing here is comparable to ``dashboard_indicator_results.color``, and
    joining the two would imply an equivalence the state does not make.

The grain is the LEA, not the school
    Districts, county offices and charter schools that are their own LEA.  A
    school inherits its district's answer; it does not have one.

The columns change almost every year
    Priority 3 has been published with 8, 21, 27 and 28 columns; Priority 6's
    single ``summary`` became ``summary1``/``summary2``/``summary3``; the key
    column has been spelled ``CDSCode``, ``cdsCode`` and ``cdscode``, and the
    priority ``PriorityNumber`` then ``priorityId``.  Only a small envelope is
    stable, so only that is given columns.  Everything else is kept verbatim in
    ``responses``, exactly as ``source_extras`` does on the state side.
"""

from __future__ import annotations

import datetime
from typing import Any

from sqlalchemy import Index
from sqlalchemy.dialects.postgresql import JSONB
from sqlmodel import Column, Field

from app.model.reference import ApiModel

#: The three values the state publishes for a local indicator.
MET = "Met"
NOT_MET = "Not Met"
NOT_MET_TWO_YEARS = "Not Met For Two or More Years"

PERFORMANCE_VALUES = (MET, NOT_MET, NOT_MET_TWO_YEARS)


class LocalIndicatorPriority(ApiModel, table=True):
    """One of the seven LCFF state priorities reported as a local indicator."""

    __tablename__ = "local_indicator_priorities"

    priority_number: int = Field(primary_key=True)
    name: str = Field(max_length=120)
    short_name: str = Field(max_length=60)
    description: str | None = Field(default=None)
    #: Priorities 9 and 10 are reported only by county offices of education.
    county_office_only: bool = Field(default=False)
    sort_order: int = Field(default=0)


class LocalIndicatorResult(ApiModel, table=True):
    """One LEA's self-assessment against one priority, for one year."""

    __tablename__ = "local_indicator_results"
    __table_args__ = (
        # The primary key leads with the entity, which serves the LEA page.
        # This covers the other direction -- one priority across many LEAs --
        # for statewide summaries.
        Index(
            "ix_local_indicator_lookup",
            "reporting_year",
            "priority_number",
            "performance",
        ),
    )

    cds_code: str = Field(
        primary_key=True, max_length=14, foreign_key="entities.cds_code"
    )
    reporting_year: int = Field(primary_key=True)
    priority_number: int = Field(
        primary_key=True, foreign_key="local_indicator_priorities.priority_number"
    )

    #: The LEA's name as it appears in the file, which is not always the name
    #: the entity carries from the assessment files.
    lea_name: str | None = Field(default=None, max_length=200)

    performance: str | None = Field(default=None, max_length=40)

    #: When the governing board received the self-assessment at a public
    #: meeting.  This is the only provenance a reader has for the narratives,
    #: so it travels with them rather than being dropped as bookkeeping.
    meeting_date: datetime.date | None = Field(default=None)

    additional_info: str | None = Field(default=None)

    #: Every column outside the stable envelope, under the name the state gave
    #: it that year.  Ratings arrive as integers, narratives as text.
    responses: dict[str, Any] = Field(
        default_factory=dict, sa_column=Column(JSONB, nullable=False, default=dict)
    )
